### ISSUE TITLE
Complicated semantics of deserializing non-empty vectors

### ISSUE THREAD

**Issue by mavam** [actor-framework/actor-framework#786]
Consider this snippet:

```cpp
struct foo {
  std::vector<int> xs = {0};
};

template <class Inspector>
auto inspect(Inspector& f, foo& x) {
  return f(x.xs);
}
```

The `inspect` function is seemingly correct but the code is buggy: a deserialized `foo` will have one extra element, because the deserialization routine assume a default-constructed (= empty) vector. This implementation works, however:

```cpp
template <class Inspector>
auto inspect(Inspector& f, foo& x) {
  if constexpr (Inspector::writes_state)
    x.xs.clear();
  return f(x.xs);
}
```

But it feels wrong to me. I feel that it's up to the framework to call `clear`. What do you think?

**Comment by Neverlord** [comment]
I don't have strong feelings towards a change. However, why would you ever implement a default-constructor that pre-fills a member? Just curious.

**Comment by mavam** [comment]
> [..] why would you ever implement a default-constructor that pre-fills a member? Just curious.

The scenario is that an empty vector does not make sense. In my use case, I implemented a hasher that produces one or more hash functions. 0 hash functions would render it dysfunctional. In order for it to have a default-constructed state that doesn't violate invariants, I start with a one-element vector.

**Comment by Neverlord** [comment]
Should be fixed now via e9398439d22ad7c6f75e86338d49252b8518a903.
